core/extensions/impl/storage_extension.cpp

```diff
@@ -131,9 +131,12 @@ namespace kagome::extensions {
 
     auto key = memory_->loadN(key_ptr, key_size);
     boost::optional<uint32_t> res{boost::none};
-    if (auto data = get(key, offset, value_size)) {
-      memory_->storeBuffer(value_ptr, data.value());
-      res = value_size;
+    if (const auto& data_res = get(key); data_res) {
+      const auto& data = data_res.value();
+      auto offset_data = data.subbuffer(std::min<size_t>(offset, data.size()));
+      auto written = std::min<size_t>(offset_data.size(), value_size);
+      memory_->storeBuffer(value_ptr, offset_data.resize(written));
+      res = offset_data.size();
```

Comment:
I think it would be more optimal to first create a span over data and then shrink it accroding to the offset and the out buffer size, and then make a buffer out of it and store it in wasm memory. Because for now as I can see it is first partially copied to offset_data, then partially copied when resized.